src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+            // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
+            Ok(obj.into_py(unsafe { Python::assume_gil_acquired() }))
+        };
+        Self::new(Box::pin(wrap), None)
+    }
+
+    fn poll(
+        &mut self,
+        py: Python<'_>,
+        throw: Option<&PyAny>,
+    ) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        // raise if the coroutine has already been run to completion
+        let future_rs = match self.future {
+            Some(ref mut fut) => fut,
+            None => return Err(PyRuntimeError::new_err(COROUTINE_REUSED_ERROR)),
+        };
+        // pass thrown exception to throw callback if there is one, otherwise reraise it
+        if let Some(exc) = throw {
+            if let Some(cb) = &mut self.throw_callback {
+                cb(exc)
+            } else {
+                self.future.take();
+                return Err(PyErr::from_value(exc));
+            };
+        }
+        loop {
+            // create a waker from an `asyncio.Future`, or try to modify it in place
+            if let Some(waker) = self.waker.as_mut().and_then(Arc::get_mut) {
+                waker.future = asyncio_future(py)?;
+            } else {
+                self.waker = Some(Arc::new(PyFutureWaker {
+                    future: asyncio_future(py)?,
+                    thread_id: current_thread_id(),
+                }));
+            }
+            let waker = futures::task::waker(self.waker.clone().unwrap());
+            // poll the Rust future and forward its results if ready
+            if let Poll::Ready(res) = future_rs.as_mut().poll(&mut Context::from_waker(&waker)) {
+                self.future.take();
+                return Ok(IterNextOutput::Return(res?));
+            }
+            // `asyncio.Future` must be awaited; fortunately, it implements `__iter__ = __await__`
+            // and will yield itself if its result has not been set in polling above
+            let future = self.waker.as_ref().unwrap().future.as_ref(py);
+            if let Some(future) = PyIterator::from_object(future).unwrap().next() {
+                // future has not been leaked into Python for now, and Rust code can only call
+                // `set_result(None)` in `ArcWake` implementation, so it's safe to unwrap
+                return Ok(IterNextOutput::Yield(future.unwrap().into()));
+            }
+            // Python future was done, i.e. `FutureWaker::wake_by_ref` has been called in
+            // `future_rs.as_mut().poll(...)`, so continue looping
+        }
+    }
+}
+
+pub(crate) fn iter_result(result: IterNextOutput<PyObject, PyObject>) -> PyResult<PyObject> {
+    match result {
+        IterNextOutput::Yield(ob) => Ok(ob),
+        IterNextOutput::Return(ob) => Err(PyStopIteration::new_err(ob)),
+    }
+}
+
+#[pymethods(crate = "crate")]
+impl Coroutine {
+    fn send(&mut self, py: Python<'_>, _value: &PyAny) -> PyResult<PyObject> {
+        iter_result(self.poll(py, None)?)
```

Comment:
Should we throw if `_value` is not `None`?

Reply:
As written above, Python coroutines seems to always ignore the value, so prefer to keep this behavior.
Example:
```python
import asyncio
async def main():
    async def f():
        await asyncio.sleep(1)
    coro = f()
    fut = coro.send(None)
    fut.set_result(None)
    try:
        coro.send(42)  # doesn't raise because of 42
    except StopIteration:
        print("ok")
```

Reply:
Ok a good reason to just ignore it 👍